libs/vkd3d/command.c

```diff
@@ -2079,6 +2109,202 @@ static void d3d12_command_list_clear_attachment_inline(struct d3d12_command_list
+            return;
+        }
+    }
+}
+
+static uint32_t d3d12_command_list_promote_dsv_resource(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, uint32_t plane_optimal_mask)
+{
+    size_t i, n;
+    assert(!(plane_optimal_mask & ~(VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL)));
+
+    /* No point in adding these since they are always deduced to be optimal. */
+    if (resource->desc.Flags & D3D12_RESOURCE_FLAG_DENY_SHADER_RESOURCE)
+        return VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL;
+
+    if (!(resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_STENCIL_BIT))
+        plane_optimal_mask |= (plane_optimal_mask & VKD3D_DEPTH_PLANE_OPTIMAL) ? VKD3D_STENCIL_PLANE_OPTIMAL : 0;
+    if (!(resource->format->vk_aspect_mask & VK_IMAGE_ASPECT_DEPTH_BIT))
+        plane_optimal_mask |= (plane_optimal_mask & VKD3D_STENCIL_PLANE_OPTIMAL) ? VKD3D_DEPTH_PLANE_OPTIMAL : 0;
+
+    for (i = 0, n = list->dsv_resource_tracking_count; i < n; i++)
+    {
+        if (list->dsv_resource_tracking[i].resource == resource)
+        {
+            list->dsv_resource_tracking[i].plane_optimal_mask |= plane_optimal_mask;
+            return list->dsv_resource_tracking[i].plane_optimal_mask;
+        }
+    }
+
+    vkd3d_array_reserve((void **)&list->dsv_resource_tracking, &list->dsv_resource_tracking_size,
+            list->dsv_resource_tracking_count + 1, sizeof(*list->dsv_resource_tracking));
+    list->dsv_resource_tracking[list->dsv_resource_tracking_count].resource = resource;
+    list->dsv_resource_tracking[list->dsv_resource_tracking_count].plane_optimal_mask = plane_optimal_mask;
+    list->dsv_resource_tracking_count++;
+    return plane_optimal_mask;
+}
+
+static uint32_t d3d12_command_list_notify_dsv_writes(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, const struct vkd3d_view *view, uint32_t plane_write_mask)
+{
+    assert(!(plane_write_mask & ~(VKD3D_DEPTH_PLANE_OPTIMAL | VKD3D_STENCIL_PLANE_OPTIMAL)));
+
+    /* If we cover the entire resource, we can promote it to our target layout. */
+    if (view->info.texture.layer_count == resource->desc.DepthOrArraySize &&
+            resource->desc.MipLevels == 1)
+    {
+        return d3d12_command_list_promote_dsv_resource(list, resource, plane_write_mask);
+    }
+    else
+    {
+        d3d12_command_list_get_depth_stencil_resource_layout(list, resource, &plane_write_mask);
+        return plane_write_mask;
+    }
+}
+
+static void d3d12_command_list_notify_dsv_state(struct d3d12_command_list *list,
+        struct d3d12_resource *resource, D3D12_RESOURCE_STATES state, UINT subresource)
+{
+    /* Need to decide if we should promote or decay or promote DSV optimal state.
+     * We can promote if we know for sure that all subresources are optimal.
+     * If we observe any barrier which leaves this state, we must decay. */
+    bool dsv_optimal = state == D3D12_RESOURCE_STATE_DEPTH_READ ||
```

Comment:
Do we know of any broken apps trying to read a `D3D12_RESOURCE_STATE_DEPTH_READ` resource inside shaders without also specifying a shader resource state?

I hope I'm wrong but this smells like it'll need app workarounds in the future.

Reply:
I don't know of any case. Worst case, we can pretend DEPTH_READ is not an optimal-type layout.

Reply:
Added comments.